stempy/reader.cpp

```diff
@@ -103,15 +139,26 @@ Header StreamReader::readHeaderVersion2() {
   return header;
 }
 
-Block StreamReader::read() {
+Block& StreamReader::read()
```

Comment:
Why the change in return type? The ownership semantic aren't clear here. Before the caller took ownership, here I'm not so sure? It seems it might always be a reference to `m_curBlock`, which would then change as the iteration continues, which is not what we want?

Reply:
This was done for the `StreamReader::iterator`.

For input iterators, `++it` is supposed to increment the iterator and then return the iterator. In our case, the increment is done by calling `StreamReader::read()`. If we don't keep a copy of the current block inside the `StreamReader` class, then the block will be destroyed immediately after creation when `++it` is called. We also need to store a copy of the current block inside the `StreamReader` class so that `*it` will return the current block. 

So now, the `StreamReader` class has ownership of the current block, and it returns references when possible to avoid making too many copies. However, multiple copies of a `block` can be made (multiple ownership), and the pointer will be deleted when they all get destroyed (thanks to `shared_ptr`). If we want to take ownership of the block from `StreamReader`, we can now use `std::move` on either the return of `read()` or the return of `*it`.

Reply:
Just so I understand. In the case that the using does the following:

```
Block& blockA = reader.read();
Block& blockB = reader.read();
```

Do `blockA` and 'blockB' now point to the same block?

Reply:
Yeah, I think they do.

I was thinking about it some last night. If you think it'd be better to return `Block` from `read()`, we can make it so that `read()` doesn't set `m_curBlock`. But we can add another function or something called `increment` or something, that calls `read()` and sets `m_curBlock`. Then the `StreamReader::iterator` can use that function instead of `read()`. Just an idea.

Reply:
Too me that is a little confusing, lets chat about this.

Reply:
This sounds like it is going to be a possible source of subtle bugs, and we should make the iterator store the current block, not overload the read method. Just catching up on some of what has been happening, but this doesn't look intuitive in its current state.

Reply:
Yeah, I'm fixing the `read()` to return a `Block`, and moving the current block into the iterator now. We started looking at `std::istream_iterator` as a reference, as it seems fairly similar to what we are doing.